airbyte-cdk/python/airbyte_cdk/sources/file_based/file_types/csv_parser.py

```diff
@@ -129,96 +193,199 @@ def _to_nullable(row: Mapping[str, str], null_values: Set[str]) -> Dict[str, Opt
         return nullable
 
     @staticmethod
-    def _skip_rows_before_header(fp: IOBase, rows_to_skip: int) -> None:
-        """
-        Skip rows before the header. This has to be done on the file object itself, not the reader
+    def _cast_types(
+        row: Dict[str, str], property_types: Dict[str, Any], config_format: CsvFormat, logger: logging.Logger
+    ) -> Dict[str, Any]:
         """
-        for _ in range(rows_to_skip):
-            fp.readline()
+        Casts the values in the input 'row' dictionary according to the types defined in the JSON schema.
 
-    def _get_headers(self, fp: IOBase, config_format: CsvFormat, dialect_name: str) -> List[str]:
-        # Note that this method assumes the dialect has already been registered if we're parsing the headers
-        if config_format.autogenerate_column_names:
-            return self._auto_generate_headers(fp, config_format)
-        else:
-            # If we're not autogenerating column names, we need to skip the rows before the header
-            self._skip_rows_before_header(fp, config_format.skip_rows_before_header)
-            # Then read the header
-            reader = csv.DictReader(fp, dialect=dialect_name)  # type: ignore
-            return next(reader)  # type: ignore
+        Array and object types are only handled if they can be deserialized as JSON.
 
-    def _auto_generate_headers(self, fp: IOBase, config_format: CsvFormat) -> List[str]:
-        """
-        Generates field names as [f0, f1, ...] in the same way as pyarrow's csv reader with autogenerate_column_names=True.
-        See https://arrow.apache.org/docs/python/generated/pyarrow.csv.ReadOptions.html
+        If any errors are encountered, the value will be emitted as a string.
         """
-        next_line = next(fp).strip()
-        number_of_columns = len(next_line.split(config_format.delimiter))  # type: ignore
-        # Reset the file pointer to the beginning of the file so that the first row is not skipped
-        fp.seek(0)
-        return [f"f{i}" for i in range(number_of_columns)]
+        warnings = []
+        result = {}
+
+        for key, value in row.items():
+            prop_type = property_types.get(key)
+            cast_value: Any = value
+
+            if isinstance(prop_type, list):
+                prop_type_distinct = set(prop_type)
+                prop_type_distinct.remove("null")
+                if len(prop_type_distinct) != 1:
+                    raise ValueError(f"Could not get non nullable type from {prop_type}")
+                (prop_type,) = prop_type_distinct
+
+            if prop_type in TYPE_PYTHON_MAPPING and prop_type is not None:
```

Comment:
snippet from below:
```
TYPE_PYTHON_MAPPING: Mapping[str, Tuple[str, Optional[Type[Any]]]] = {
    "null": ("null", None),
    "array": ("array", list),
    "boolean": ("boolean", bool),
    "float": ("number", float),
    "integer": ("integer", int),
    "number": ("number", float),
    "object": ("object", dict),
    "string": ("string", str),
}
PYTHON_TYPE_MAPPING = {t: k for k, (_, t) in TYPE_PYTHON_MAPPING.items()}
```

Given that `TYPE_PYTHON_MAPPING` is all str -> tuples, I'm not sure when `prop_type` would resolve to a valid key while also being `None`. Do we need this second condition?

I could see that being the case if we were using the PYTHON_TYPE_MAPPING since this could yield None -> "null"

Reply:
You are right. This is probably legacy from the previous code. I'll remove it